### PR TITLE
[WIP] Metropolis: EIP86 account abstraction

### PR DESCRIPTION
**Pull request by pirapira**
This PR introduces https://github.com/ethereum/EIPs/pull/208 as part of the Metropolis changes.  #229 

This is a reincarnation of #249, but reflects a recent change in https://github.com/ethereum/EIPs/pull/208.

### CODE REVIEW

File: Paper.tex [hunk 1]
```diff
@@ -594,13 +594,17 @@ \subsection{Execution}
 \begin{eqnarray}
 \boldsymbol{\sigma}_0 & \equiv & \boldsymbol{\sigma} \quad \text{except:} \\
 \boldsymbol{\sigma}_0[S(T)]_b & \equiv & \boldsymbol{\sigma}[S(T)]_b - T_g T_p \\
-\boldsymbol{\sigma}_0[S(T)]_n & \equiv & \boldsymbol{\sigma}[S(T)]_n + 1
+\boldsymbol{\sigma}_0[S(T)]_n & \equiv &
+\begin{cases}
+\boldsymbol{\sigma}[S(T)]_n &\text{if}\ S(T) = 2^{256} \\
```
Comment (pirapira) [inline Paper.tex]: I guess the number should be `2^{256} - 1`.

File: Paper.tex [hunk 2]
```diff
@@ -609,7 +613,11 @@ \subsection{Execution}
 \begin{equation}
 g \equiv T_g - g_0
 \end{equation}
-and $T_o$ is the original transactor, which can differ from the sender in the case of a message call or contract creation not directly triggered by a transaction but coming from the execution of EVM-code.
+and $T_o$ is the original transactor, which can differ from the sender in the case of a message call or contract creation not directly triggered by a transaction but coming from the execution of EVM-code.  $a$ is the address of the created account:
```
Comment (gavofyork) [inline Paper.tex]: superfluous double-space

File: Paper.tex [hunk 3]
```diff
@@ -609,7 +613,11 @@ \subsection{Execution}
...
+\begin{equation} \label{eq:address_calc}
+a \equiv \mathcal{B}_{96..255}\Big(\mathtt{\tiny KEC}\big(0 \cdot \mathbf{i}\big)\Big)
+\end{equation}
+where $0$ denotes the zero element of $\mathbb{B}_{160}$, the operator $\cdot$ is the concatenation of binary data, $\mathtt{\tiny KEC}$ is the Keccak 256-bit hash function, $\mathcal{B}_{a..b}(X)$ evaluates to binary value containing the bits of indices in the range $[a, b]$ of the binary data.
```
Comment (gavofyork) [inline Paper.tex]: better not to "denote" `0` to be anything other than zero. if you want to pun it into a particular element of a set, better to use some other notation.

File: Paper.tex (hunk 1 above)
Comment (gavofyork) [inline Paper.tex]: agree

File: Paper.tex [hunk 4]
```diff
@@ -609,7 +613,11 @@ \subsection{Execution}
 \begin{equation}
 g \equiv T_g - g_0
 \end{equation}
-and $T_o$ is the original transactor, which can differ from the sender in the case of a message call or contract creation not directly triggered by a transaction but coming from the execution of EVM-code.
+and $T_o$ is the original transactor, which can differ from the sender in the case of a message call or contract creation not directly triggered by a transaction but coming from the execution of EVM-code.  $a$ is the address of the created account:
+\begin{equation} \label{eq:address_calc}
+a \equiv \mathcal{B}_{96..255}\Big(\mathtt{\tiny KEC}\big(0 \cdot \mathbf{i}\big)\Big)
```
Comment (gavofyork) [inline Paper.tex]: should read KEC(\mathbf{i})

File: Paper.tex [hunk 5]
```diff
@@ -2022,6 +2031,11 @@ \subsection{Instruction Set}
 &&&& This means that the recipient is in fact the same account as at present, simply\\
 &&&& that the code is overwritten {\it and} the context is almost entirely identical.\\
 \midrule
+0xfb & {\small CREATE\_P2SH} & 3 & 1 & Create a new account with associated code on an address calculated using the \\
+&&&& creator's address. \\
+&&&& Exactly equivalent to {\small CREATE} except: \\
+&&&& $a \equiv \mathcal{B}_{96..255}\Big(\mathtt{\tiny KEC}\big(I_a \cdot \mathbf{i}\big)\Big)$ \\
```
Comment (gavofyork) [inline Paper.tex]: should read `...KEC(\mathbf{i})...`

File: Paper.tex (hunk 1 above)
Comment (pirapira) [inline Paper.tex]: I fixed this.

File: Paper.tex (hunk 2 above)
Comment (pirapira) [inline Paper.tex]: removed.

File: Paper.tex (hunk 4 above)
Comment (pirapira) [inline Paper.tex]: fixed.

File: Paper.tex (hunk 3 above)
Comment (pirapira) [inline Paper.tex]: Introduced a symbol.

File: Paper.tex (hunk 5 above)
Comment (pirapira) [inline Paper.tex]: fixed.

**Comment by pirapira** [comment]
Addressed the comments.

**Comment by pirapira** [comment]
Probably there will be a conflict with #278.

**Comment by pirapira** [comment]
This has to be adjusted after #232 and #303 are merged.

File: Paper.tex [hunk 6]
```diff
@@ -711,16 +711,17 @@ \section{Contract Creation} \label{ch:create}
...
 \end{cases} \\
 \quad \boldsymbol{\sigma}' &\equiv  \begin{cases}
-\boldsymbol{\sigma} & \text{if} \quad \boldsymbol{\sigma}^{**} = \varnothing \\
+\boldsymbol{\sigma} & \text{if} \quad \boldsymbol{\sigma}^{**} = \varnothing \vee E(\boldsymbol{\sigma}, a) \\
 \boldsymbol{\sigma}^{**} & \text{if} \quad g^{**}<c \wedge H_i<\firsthomesteadblock \\
 \boldsymbol{\sigma}^{**} \quad \text{except:} & \\
 \quad\boldsymbol{\sigma}'[a]_c = \texttt{\small KEC}(\mathbf{o}) & \text{otherwise}
-\end{cases}
+\end{cases} \\
+E(\boldsymbol{\sigma}, a) &\equiv \big(\boldsymbol{\sigma}[a] \neq \varnothing \wedge (\boldsymbol{\sigma}[a]_c \neq () \vee \boldsymbol{\sigma}[a]_n \neq 0)\big)
```
Comment (pirapira) [inline Paper.tex]: The empty code is seen as `KEC(())` not just `()`.

**Comment by pirapira** [comment]
This PR needs to follow https://github.com/ethereum/EIPs/commit/bd136e6

**Comment by pirapira** [comment]
I finished applying ethereum/EIPs@bd136e6

**Comment by pirapira** [comment]
This protocol change has been postponed and I don't know when it is planned.

**Comment by Hektoreth** [comment]
> This PR introduces [ethereum/EIPs#208](https://github.com/ethereum/EIPs/pull/208) as part of the Metropolis changes. #229
> 
> This is a reincarnation of #249, but reflects a recent change in [ethereum/EIPs#208](https://github.com/ethereum/EIPs/pull/208).

#883

**Pull request closed without merging**